File: results/1.0/GZIP/gzip-claude-3-5-haiku-20241022.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <hammer/hammer.h>

typedef struct {
    uint8_t id1;
    uint8_t id2;
    uint8_t compression_method;
    struct {
        uint8_t text_hint : 1;
        uint8_t header_crc : 1;
        uint8_t extra_field : 1;
        uint8_t orig_filename : 1;
        uint8_t file_comment : 1;
        uint8_t reserved : 3;
    } flags;
    uint32_t modification_time;
    uint8_t extra_flags;
    uint8_t operating_system;
    
    struct {
        bool present;
        uint8_t* data;
        size_t length;
    } extra_field;
    
    struct {
        bool present;
        char* filename;
    } orig_filename;
    
    struct {
        bool present;
        char* comment;
    } file_comment;
    
    uint16_t header_crc;
    
    uint8_t* compressed_data;
    size_t compressed_length;
    
    uint32_t data_crc32;
    uint32_t original_length;
} GzipFile;

HParser* parse_gzip_header() {
    HParser* id1 = h_literal(h_ch(0x1F));
    HParser* id2 = h_literal(h_ch(0x8B));
    HParser* compression_method = h_literal(h_ch(0x08));
    
    HParser* flags = h_bits(8, false);
    HParser* modification_time = h_uint32();
    HParser* extra_flags = h_uint8();
    HParser* operating_system = h_uint8();
    
    HParser* gzip_header = h_sequence(
        id1, 
        id2, 
        compression_method, 
        flags,
        modification_time,
        extra_flags,
        operating_system,
        NULL
    );
    
    return gzip_header;
}

HParser* parse_optional_fields() {
    HParser* extra_field_length = h_uint16();
    HParser* extra_field_data = h_repeat_n(h_uint8(), 1);
    
    HParser* filename = h_many1(h_except(h_ch('\0')));
    HParser* comment = h_many1(h_except(h_ch('\0')));
    
    HParser* header_crc = h_uint16();
    
    HParser* optional_fields = h_choice(
        h_optional(extra_field_length),
        h_optional(extra_field_data),
        h_optional(filename),
        h_optional(comment),
        h_optional(header_crc),
        NULL
    );
    
    return optional_fields;
}

HParser* parse_gzip_file() {
    HParser* header = parse_gzip_header();
    HParser* optional = parse_optional_fields();
    HParser* compressed_data = h_many(h_uint8());
    HParser* data_crc32 = h_uint32();
    HParser* original_length = h_uint32();
    
    HParser* gzip_file = h_sequence(
        header,
        optional,
        compressed_data,
        data_crc32,
        original_length,
        NULL
    );
    
    return gzip_file;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <gzip_file>\n", argv[0]);
        return 1;
    }
    
    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }
    
    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);
    
    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }
    
    if (fread(buffer, 1, file_size, file) != file_size) {
        perror("File read error");
        free(buffer);
        fclose(file);
        return 1;
    }
    
    HParser* gzip_parser = parse_gzip_file();
    HParseResult* result = h_parse(gzip_parser, buffer, file_size);
    
    if (result && result->ast) {
        printf("GZIP file parsed successfully\n");
    } else {
        printf("GZIP file parsing failed\n");
    }
    
    free(buffer);
    fclose(file);
    h_parse_result_free(result);
    h_arena_free(h_get_allocator());
    
    return 0;
}

File: results/1.0/GZIP/gzip-claude-3-5-haiku-20241022.rs
use nom::{
    bytes::complete::{take_till, take_while},
    combinator::map,
    multi::many0,
    number::complete::{le_u16, le_u32, le_u8},
    sequence::tuple,
    IResult,
};
use std::env;
use std::fs;
use std::io::Read;

#[derive(Debug)]
struct GzipHeader {
    id1: u8,
    id2: u8,
    compression_method: u8,
    flags: u8,
    mod_time: u32,
    extra_flags: u8,
    operating_system: u8,
    extra_fields: Option<Vec<ExtraField>>,
    original_filename: Option<String>,
    file_comment: Option<String>,
    header_crc16: Option<u16>,
}

#[derive(Debug)]
struct ExtraField {
    subfield_id: String,
    subfield_length: u16,
    data: Vec<u8>,
}

#[derive(Debug)]
struct GzipFile {
    header: GzipHeader,
    compressed_data: Vec<u8>,
    crc32: u32,
    input_size: u32,
}

fn parse_extra_field(input: &[u8]) -> IResult<&[u8], ExtraField> {
    let (input, (subfield_id, subfield_length)) = tuple((
        map(take_while(|b: u8| b.is_ascii()), |id: &[u8]| String::from_utf8_lossy(id).into_owned()),
        le_u16,
    ))(input)?;

    let (input, data) = take_while(|_| true)(input)?;
    let data = data[..subfield_length as usize].to_vec();

    Ok((input, ExtraField {
        subfield_id,
        subfield_length,
        data,
    }))
}

fn parse_gzip_header(input: &[u8]) -> IResult<&[u8], GzipHeader> {
    let (input, (id1, id2, compression_method, flags, mod_time, extra_flags, operating_system)) = tuple((
        le_u8, le_u8, le_u8, le_u8, le_u32, le_u8, le_u8,
    ))(input)?;

    let (input, extra_fields) = if flags & 0b100 != 0 {
        let (input, fields) = many0(parse_extra_field)(input)?;
        (input, Some(fields))
    } else {
        (input, None)
    };

    let (input, original_filename) = if flags & 0b1000 != 0 {
        let (input, filename) = take_till(|b| b == 0)(input)?;
        (input, Some(String::from_utf8_lossy(filename).into_owned()))
    } else {
        (input, None)
    };

    let (input, file_comment) = if flags & 0b10000 != 0 {
        let (input, comment) = take_till(|b| b == 0)(input)?;
        (input, Some(String::from_utf8_lossy(comment).into_owned()))
    } else {
        (input, None)
    };

    let (input, header_crc16) = if flags & 0b10 != 0 {
        let (input, crc) = le_u16(input)?;
        (input, Some(crc))
    } else {
        (input, None)
    };

    Ok((input, GzipHeader {
        id1,
        id2,
        compression_method,
        flags,
        mod_time,
        extra_flags,
        operating_system,
        extra_fields,
        original_filename,
        file_comment,
        header_crc16,
    }))
}

fn parse_gzip_file(input: &[u8]) -> IResult<&[u8], GzipFile> {
    let (input, header) = parse_gzip_header(input)?;

    let (input, compressed_data) = take_while(|_| true)(input)?;
    let compressed_data = compressed_data.to_vec();

    let (input, crc32) = le_u32(input)?;
    let (input, input_size) = le_u32(input)?;

    Ok((input, GzipFile {
        header,
        compressed_data,
        crc32,
        input_size,
    }))
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <gzip_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = fs::File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_gzip_file(&buffer) {
        Ok((_, gzip_file)) => {
            println!("{:#?}", gzip_file);
            Ok(())
        }
        Err(e) => {
            eprintln!("Error parsing GZIP file: {:?}", e);
            std::process::exit(1);
        }
    }
}